From 686662c6f15d52318821d449787eea75432940f6 Mon Sep 17 00:00:00 2001
From: Binyuan Lan <lby@rock-chips.com>
Date: Sat, 10 Mar 2018 10:05:40 +0800
Subject: [PATCH] ASoC: rockchip: i2s: protect I2S_XFER_TXS/I2S_XFER_RXS with
 mutex

fix bug error log: "rockchip-i2s ff070000.i2s: fail to clear".
When the TX/RX have started, can not do 'clear operation'.

Change-Id: I83c2ac29412a37a5de02ed9d4be62716fe46678a
Signed-off-by: Binyuan Lan <lby@rock-chips.com>
---
 sound/soc/rockchip/rockchip_i2s.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/sound/soc/rockchip/rockchip_i2s.c b/sound/soc/rockchip/rockchip_i2s.c
index 7687368779db..650f6b5200ab 100644
--- a/sound/soc/rockchip/rockchip_i2s.c
+++ b/sound/soc/rockchip/rockchip_i2s.c
@@ -32,6 +32,7 @@ struct rk_i2s_pins {
 
 struct rk_i2s_dev {
 	struct device *dev;
+	struct mutex lock; /* txctrl/rxctrl lock */
 
 	struct clk *hclk;
 	struct clk *mclk;
@@ -95,6 +96,7 @@ static void rockchip_snd_txctrl(struct rk_i2s_dev *i2s, int on)
 	unsigned int val = 0;
 	int retry = 10;
 
+	mutex_lock(&i2s->lock);
 	if (on) {
 		regmap_update_bits(i2s->regmap, I2S_DMACR,
 				   I2S_DMACR_TDE_ENABLE, I2S_DMACR_TDE_ENABLE);
@@ -129,12 +131,14 @@ static void rockchip_snd_txctrl(struct rk_i2s_dev *i2s, int on)
 				regmap_read(i2s->regmap, I2S_CLR, &val);
 				retry--;
 				if (!retry) {
-					dev_warn(i2s->dev, "fail to clear\n");
+					dev_warn(i2s->dev, "%s: fail to clear\n",
+						 __func__);
 					break;
 				}
 			}
 		}
 	}
+	mutex_unlock(&i2s->lock);
 }
 
 static void rockchip_snd_rxctrl(struct rk_i2s_dev *i2s, int on)
@@ -142,6 +146,7 @@ static void rockchip_snd_rxctrl(struct rk_i2s_dev *i2s, int on)
 	unsigned int val = 0;
 	int retry = 10;
 
+	mutex_lock(&i2s->lock);
 	if (on) {
 		regmap_update_bits(i2s->regmap, I2S_DMACR,
 				   I2S_DMACR_RDE_ENABLE, I2S_DMACR_RDE_ENABLE);
@@ -176,12 +181,14 @@ static void rockchip_snd_rxctrl(struct rk_i2s_dev *i2s, int on)
 				regmap_read(i2s->regmap, I2S_CLR, &val);
 				retry--;
 				if (!retry) {
-					dev_warn(i2s->dev, "fail to clear\n");
+					dev_warn(i2s->dev, "%s: fail to clear\n",
+						 __func__);
 					break;
 				}
 			}
 		}
 	}
+	mutex_unlock(&i2s->lock);
 }
 
 static int rockchip_i2s_set_fmt(struct snd_soc_dai *cpu_dai,
@@ -601,6 +608,8 @@ static int rockchip_i2s_probe(struct platform_device *pdev)
 
 	i2s->dev = &pdev->dev;
 
+	mutex_init(&i2s->lock);
+
 	i2s->grf = syscon_regmap_lookup_by_phandle(node, "rockchip,grf");
 	if (!IS_ERR(i2s->grf)) {
 		of_id = of_match_device(rockchip_i2s_match, &pdev->dev);
-- 
2.35.3

